lib/behavior.js

```diff
@@ -0,0 +1,170 @@
+            );
+          }
+        );
+      } else {
+        if (this.asyncOnly) {
+          return done(
+            new Error(
+              '--async-only option in use without declaring `done()` or returning a promise'
+            )
+          );
+        }
+        done();
+      }
+    },
+    /**
+     * Runs the Runnable, passing a nodeback function, which must be called to
+     * complete the test
+     * @param {Function} done - Callback
+     */
+    runWithCallback: function(done) {
+      var result = this.fn.call(this.ctx, function(err) {
+        if (
+          err instanceof Error ||
+          Object.prototype.toString.call(err) === '[object Error]'
+        ) {
+          return done(err);
+        }
+        if (err) {
+          if (Object.prototype.toString.call(err) === '[object Object]') {
+            return done(
+              new Error('done() invoked with non-Error: ' + JSON.stringify(err))
+            );
+          }
+          return done(new Error('done() invoked with non-Error: ' + err));
+        }
+        if (result && utils.isPromise(result)) {
+          return done(
+            new Error(
+              'Resolution method is overspecified. Specify a callback *or* return a Promise; not both.'
+            )
+          );
+        }
+        done();
+      });
+    }
+  }
+};
+
+/**
+ * Provides a test context in a functional manner for use with
+ * lambdas/arrow functions.
+ * All Runnables must either return a promise or call `runnable.done()`, where
+ * `runnable` is the first parameter to the Runnable's callback (`fn`)
+ */
+exports.Functional = {
+  Suite: {
+    /**
+     * Runs the Suite.  Calls its `fn` with NO context and the suite itself
+     * as the first parameter.
+     * @this {Suite}
+     */
+    run: function(opts) {
+      this.fn.call(null, this);
+    }
+  },
+  Runnable: {
+    /**
+     * Determines whether or not we should provide a nodeback parameter and
+     * expect it to be called; always false
+     * @this {Runnable}
+     * @returns false
+     */
+    shouldRunWithCallback: function() {
+      return false;
+    },
+
+    /**
+     * Runs the Runnable expecting a call to ctx.done() or a Promise
+     * @param {Function} done - Callback
+     */
+    run: function(done) {
+      this.ctx.done = function(err) {
+        if (
+          err instanceof Error ||
+          Object.prototype.toString.call(err) === '[object Error]'
+        ) {
+          return done(err);
+        }
+        if (err) {
+          if (Object.prototype.toString.call(err) === '[object Object]') {
+            return done(
+              new Error('done() invoked with non-Error: ' + JSON.stringify(err))
+            );
+          }
+          return done(new Error('done() invoked with non-Error: ' + err));
+        }
+        done();
+      };
+      var result = this.fn.call(null, this.ctx);
+      if (result && typeof result.then === 'function') {
+        this.resetTimeout();
```

Comment:
I think this line should ve moved above `if` statement.
Because tests just exit if `test.done()` is not called if synchronous tests. Also we should fix [the error message](https://github.com/mochajs/mocha/blob/5fc58458395d8a9bc75daae97042af78dc94cc7a/lib/runnable.js#L436) for func ui.